key_value_map.py

```diff
@@ -0,0 +1,105 @@
+from cartesian_coordinate import CartesianCoordinate
+from mapping_interface import MapInterface
+from mapping_interface import PointState
+
+
+class keyValueMap(MapInterface):
+    """
+    @brief Class keyValueMap provides functionality to create a map using keys and values.
+    """
+
+    def __init__(self):
+        """
+        @brief keyValueMap constructor used to create a basic key value map.
+        """
+
+        self.map = {}
+        self.top_left_point = CartesianCoordinate(0, 0)
+        self.bot_right_point = CartesianCoordinate(0, 0)
+
+    def get_top_left_point(self):
+        """
+        @brief Method used to get the top_left_point of the key value map.
+        @return A CartesianCoordinate containing the top left corner point of the key value map.
+        """
+
+        return self.top_left_point
+
+    def get_bot_right_point(self):
+        """
+        @brief Method used to get the bot_right_point of the key value map.
+        @return A CartesianCoordinate containing the bottom right corner point of the key value map.
+        """
+
+        return self.bot_right_point
+
+    def get_height(self):
+        """
+        @brief Method used to get the height key value of the map.
+        @return A integer containing the height key value of the map.
+        """
+
+        return self.get_bot_right_point().y - self.get_top_left_point().y + 1
+
+    def get_width(self):
+        """
+        @brief Method used to get the width of the key value map.
+        @return A integer containing the width of the key value map.
+        """
+
+        return self.get_bot_right_point().x - self.get_top_left_point().x + 1
+
+    def get_map_points(self):
+        """
+        @brief Method used to get all the points stored in the map.
+        @return A list containing all the occupied points of the map.
+        """
+
+        return self.map.keys()
+
+    def is_occupied(self, coordinate):
+        """
+        @brief Checks whether a point is occupied in the key value map. 
+        @return A boolean containing whether the point is occupied.
+        """
+
+        return CartesianCoordinate(coordinate.x, coordinate.y) in self.get_map_points()
+
+    def update_top_left_point(self, new_point):
+        """
+        @brief Method used to update the TopLeftPoint.
+        @param new_point The point you want to check if it should be the new TopLeftPoint
+        @details Updates the TopLeftPoint if the new_point is outside the outer border of the current map on the top left.
+        """
+
+        if new_point.x < self.top_left_point.x:
```

Comment:
If I add a point with a larger `x` and `y`, will the top left point only scale on `y` due to this if statement?

Reply:
I dont understand the problem, i think it works fine, if the x is bigger the x will be modified first, then later the y

Reply:
Its top left points, larger in this scope is further from (0,0) which is a smaller number.

Reply:
@LRstudentHU it will work just fine, these are 'if' not 'elif' meaning that if your `x` and your `y` are larger, the `y` will use the `self.top_left_point.x` which will have been updated by the if statement at line 75.